types/wlr_scene.c

```diff
@@ -105,16 +163,88 @@ struct wlr_scene_surface *wlr_scene_surface_create(struct wlr_scene_node *parent
 	scene_surface->surface_destroy.notify = scene_surface_handle_surface_destroy;
 	wl_signal_add(&surface->events.destroy, &scene_surface->surface_destroy);
 
+	scene_surface->surface_commit.notify = scene_surface_handle_surface_commit;
+	wl_signal_add(&surface->events.commit, &scene_surface->surface_commit);
+
+	scene_node_damage_whole(&scene_surface->node);
+
 	return scene_surface;
 }
 
+static int scale_length(int length, int offset, float scale) {
+	return round((offset + length) * scale) - round(offset * scale);
+}
+
+static void scale_box(struct wlr_box *box, float scale) {
+	box->width = scale_length(box->width, box->x, scale);
+	box->height = scale_length(box->height, box->y, scale);
+	box->x = round(box->x * scale);
+	box->y = round(box->y * scale);
+}
+
+static void _scene_node_damage_whole(struct wlr_scene_node *node,
+		struct wlr_scene *scene, int lx, int ly) {
+	if (!node->state.enabled) {
+		return;
+	}
+
+	struct wlr_scene_node *child;
+	wl_list_for_each(child, &node->state.children, state.link) {
+		_scene_node_damage_whole(child, scene,
+			lx + child->state.x, ly + child->state.y);
+	}
+
+	int width, height;
+	switch (node->type) {
+	case WLR_SCENE_NODE_ROOT:
+		return;
+	case WLR_SCENE_NODE_SURFACE:;
+		struct wlr_scene_surface *scene_surface = scene_surface_from_node(node);
+		width = scene_surface->surface->current.width;
+		height = scene_surface->surface->current.height;
+		break;
+	}
+
+	struct wlr_scene_output *scene_output;
+	wl_list_for_each(scene_output, &scene->outputs, link) {
+		struct wlr_box box = {
+			.x = lx - scene_output->x,
+			.y = ly - scene_output->y,
+			.width = width,
+			.height = height,
+		};
+
+		scale_box(&box, scene_output->output->scale);
+
+		wlr_output_damage_add_box(scene_output->damage, &box);
+	}
+}
+
+static void scene_node_damage_whole(struct wlr_scene_node *node) {
+	struct wlr_scene *scene = scene_node_get_root(node);
+	if (wl_list_empty(&scene->outputs)) {
+		return;
+	}
+
+	int lx, ly;
+	if (!wlr_scene_node_coords(node, &lx, &ly)) {
+		return;
+	}
+
+	_scene_node_damage_whole(node, scene, lx, ly);
+}
+
 void wlr_scene_node_set_enabled(struct wlr_scene_node *node, bool enabled) {
+	scene_node_damage_whole(node);
 	node->state.enabled = enabled;
+	scene_node_damage_whole(node);
 }
```

Comment:
This will damage the entire subtree even if the enabled state isn't changed, which I think should be avoided.

I also was very confused by the need for two `scene_node_damage_whole()` calls at first glance, perhaps that's worth a brief comment saying that one will short-circuit.

Reply:
Good point! Added an early return to the functions you've mentioned, plus place_above/place_below.